# EXPLORING COMMENTS ON HACKER NEWS DATA SET

We work with the data set which is  took from "Y Hacker News". The Hacker News [website](https://news.ycombinator.com/) publishes stories about technology. We will examine the amount of posts which is about 'Python'. And, we will explore how many comments is taken for per hour for posts.
<br>
<br>
There are two type of titles that is released on Hacker News website. These titles are 'Ask HN' and 'Show HN'. We will focus on 'Ask HN' titles. And, we will analyze the average number of comments for per hour.
<br>
Additionally, artificial intelligence is popular topic of last years. We will explore the averge number of comments for this title.

### Opening Data

In [1]:

opened_file= open("hacker_news.csv")
from csv import reader
read_file= reader(opened_file)
hn= list(read_file)
hn[:1]


[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']]

We can seperate the core data from headers.

In [2]:
headers= hn[0]
hn= hn[1:]
print(headers)
hn[:2]

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


[['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30']]

### Seperating the 'Ask HN' and 'Show HN' titles from data

In [3]:
ask_posts=[]
show_posts=[]
other_posts=[]
for row in hn:
    title=row[1]
    if (title.lower()).startswith('ask hn'):
        ask_posts.append(row)
    if (title.lower()).startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
        
len_ask= len(ask_posts)
len_show= len(show_posts)
len_other= len(other_posts)

print(len_ask)
print(len_show)
print(len_other)

1744
1162
18938


In [4]:
ask_posts[:2]


[['12296411',
  'Ask HN: How to improve my personal website?',
  '',
  '2',
  '6',
  'ahmedbaracat',
  '8/16/2016 9:55'],
 ['10610020',
  'Ask HN: Am I the only one outraged by Twitter shutting down share counts?',
  '',
  '28',
  '29',
  'tkfx',
  '11/22/2015 13:43']]

In [23]:
show_posts[:2]

[['10627194',
  'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform',
  'https://iot.seeed.cc',
  '26',
  '22',
  'kfihihc',
  '11/25/2015 14:03'],
 ['10646440',
  'Show HN: Something pointless I made',
  'http://dn.ht/picklecat/',
  '747',
  '102',
  'dhotson',
  '11/29/2015 22:46']]

In [6]:
#To calculate the average number of comments for 'Ask HN' titles.
total_ask_comments=0
for row in ask_posts:
    num_comments= int(row[4])
    total_ask_comments+= num_comments
    avg_ask_comments= total_ask_comments/len(ask_posts)
    
print(avg_ask_comments)
    

14.038417431192661


In [7]:
#To calculate the average number of comments for 'Show HN'.
total_show_comments=0
for row in show_posts:
    num_comments=int(row[4])
    total_show_comments+= num_comments
    avg_show_comments= total_show_comments/ len(show_posts)

print(avg_show_comments)
    

    

10.31669535283993


The 'Ask HN' titles amount is more than The 'Show HN' titles amount. 
<br>
The askhn amount: 1744
<br>
The showhn amount:1162
<br>
Also the average of comments for the 'Ask HN' is more than the 'Show HN's average.
<br>
The average of comments for the 'Ask HN': 14.04
<br>
The average of comments for the 'Show HN': 10.32

We won't work with the 'Show HN' titles, anymore. Now, it is time to find the 'Ask HN' titles about Python.

### Posts About Artificial Intelligence 

In [8]:
#Determine a function to get 'Artificial Intelligence' words from titles.

def get_word_ai(text):
    sentences = text.split(".")
    for sentence in sentences:
        if "ai" in sentence.lower():
            return True
        elif "artificial intelligence" in sentence.lower():
                return True




In [9]:
#To find the number of posts about artificial intelligence in hacker news data set.
ai_posts=[]
for new in hn:
    title= new[1]
    if get_word_ai(title)==True:
        ai_posts.append(new)
        
        
print(len(ai_posts))

1842


In [10]:
#To find the number of posts about artificial intelligence in Ask HN titles.
ai_askhn_posts=[]
for new in ask_posts:
    title= new[1]
    if get_word_ai(title)==True:
        ai_askhn_posts.append(new)
        
        
print(len(ai_askhn_posts))

121


There are 1842  AI titles in the Hacker News website, 121 of them are under Ask HN titles.

### Calculating Average Amount of Comments For 'Ask HN' Titles

In [11]:
import datetime as dt

For first step, we need to obtain time and the number of comments.

In [21]:
result_list=[]
for row in ask_posts: #Create a list of time and number of comments.
    f_element= row[6]
    s_element= int(row[4])
    aa=[f_element, s_element]
    result_list.append(aa)
    
result_list[:5]
    

[['8/16/2016 9:55', 6],
 ['11/22/2015 13:43', 29],
 ['5/2/2016 10:14', 1],
 ['8/2/2016 14:20', 3],
 ['10/15/2015 16:38', 17]]

In [13]:
#We need lists that show the number of posts and the number of comment for per hour.
counts_by_hour={}
comments_by_hour={}
date_format= ("%m/%d/%Y %H:%M")
for row in result_list:
    f_element= row[0]
    s_element= row[1]
    time= dt.datetime.strptime(f_element,date_format).strftime("%H")
    if time not in counts_by_hour:
        counts_by_hour[time]=1
        comments_by_hour[time]=s_element
    else:
        counts_by_hour[time]+=1
        comments_by_hour[time]+= s_element
        
    

In [22]:
counts_by_hour #This list shows the number of posts on 'Ask HN' title for per hour.

{'09': 45,
 '13': 85,
 '10': 59,
 '14': 107,
 '16': 108,
 '23': 68,
 '12': 73,
 '17': 100,
 '15': 116,
 '21': 109,
 '20': 80,
 '02': 58,
 '18': 109,
 '03': 54,
 '05': 46,
 '19': 110,
 '01': 60,
 '22': 71,
 '08': 48,
 '04': 47,
 '00': 55,
 '06': 44,
 '07': 34,
 '11': 58}

In [24]:
comments_by_hour #This list shows the number of comments on 'Ask HN' title for per hour.

{'09': 251,
 '13': 1253,
 '10': 793,
 '14': 1416,
 '16': 1814,
 '23': 543,
 '12': 687,
 '17': 1146,
 '15': 4477,
 '21': 1745,
 '20': 1722,
 '02': 1381,
 '18': 1439,
 '03': 421,
 '05': 464,
 '19': 1188,
 '01': 683,
 '22': 479,
 '08': 492,
 '04': 337,
 '00': 447,
 '06': 397,
 '07': 267,
 '11': 641}

In [25]:
#With those of the lists, we can calculate the average number of comments.
avg_by_hour=[]
for hr in comments_by_hour:
    avg_by_hour.append([hr, (comments_by_hour[hr]/counts_by_hour[hr])])
    
avg_by_hour
                        

                            

[['09', 5.5777777777777775],
 ['13', 14.741176470588234],
 ['10', 13.440677966101696],
 ['14', 13.233644859813085],
 ['16', 16.796296296296298],
 ['23', 7.985294117647059],
 ['12', 9.41095890410959],
 ['17', 11.46],
 ['15', 38.5948275862069],
 ['21', 16.009174311926607],
 ['20', 21.525],
 ['02', 23.810344827586206],
 ['18', 13.20183486238532],
 ['03', 7.796296296296297],
 ['05', 10.08695652173913],
 ['19', 10.8],
 ['01', 11.383333333333333],
 ['22', 6.746478873239437],
 ['08', 10.25],
 ['04', 7.170212765957447],
 ['00', 8.127272727272727],
 ['06', 9.022727272727273],
 ['07', 7.852941176470588],
 ['11', 11.051724137931034]]

In [26]:
swap_avg_by_hour=[]

for row in avg_by_hour:#Change the locations of rows
    swap_avg_by_hour.append([row[1],row[0]])
    
swap_avg_by_hour
    
    

[[5.5777777777777775, '09'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [16.796296296296298, '16'],
 [7.985294117647059, '23'],
 [9.41095890410959, '12'],
 [11.46, '17'],
 [38.5948275862069, '15'],
 [16.009174311926607, '21'],
 [21.525, '20'],
 [23.810344827586206, '02'],
 [13.20183486238532, '18'],
 [7.796296296296297, '03'],
 [10.08695652173913, '05'],
 [10.8, '19'],
 [11.383333333333333, '01'],
 [6.746478873239437, '22'],
 [10.25, '08'],
 [7.170212765957447, '04'],
 [8.127272727272727, '00'],
 [9.022727272727273, '06'],
 [7.852941176470588, '07'],
 [11.051724137931034, '11']]

In [27]:
sorted_swap= sorted(swap_avg_by_hour, reverse=True)# sort the rows

sorted_swap

[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [19]:
print("Top 5 Hours for Ask Posts Comments")
for avg,hr in sorted_swap[:5]:
    print(
       " {} {:.2f} average comments per post.".format(dt.datetime.strptime(hr, "%H").strftime("%H:%M"),avg))

Top 5 Hours for Ask Posts Comments
 15:00 38.59 average comments per post.
 02:00 23.81 average comments per post.
 20:00 21.52 average comments per post.
 16:00 16.80 average comments per post.
 21:00 16.01 average comments per post.


We can see the top hours that the users make comments under 'Ask HN' titles.